cGame_logic.cpp

```diff
@@ -547,38 +554,41 @@ void cGame::shakeScreenAndBlitBuffer() {
     if (TIMER_shake == 0) {
 		TIMER_shake = -1;
 	}
-	// blitSprite on screen
 
+	// blitSprite on screen
 	if (TIMER_shake > 0)
 	{
 		// the more we get to the 'end' the less we 'throttle'.
 		// Structure explosions are 6 time units per cell.
 		// Max is 9 cells (9*6=54)
 		// the max border is then 9. So, we do time / 6
-		if (TIMER_shake > 69) TIMER_shake = 69;
-
-		int offset = TIMER_shake / 5;
-		if (offset > 9)
-			offset = 9;
+    TIMER_shake = std::min(TIMER_shake, 69);
+    int offset = std::min(TIMER_shake / 5, 9);
 
 		shake_x = -abs(offset/2) + rnd(offset);
 		shake_y = -abs(offset/2) + rnd(offset);
 
-		blit(bmp_screen, bmp_throttle, 0, 0, 0+shake_x, 0+shake_y, screen_x, screen_y);
+		blit(bmp_screen, bmp_throttle, 0, 0, 0 + shake_x, 0 + shake_y, screen_x, screen_y);
+    vsync();
 		blit(bmp_throttle, screen, 0, 0, 0, 0, screen_x, screen_y);
 	}
 	else
 	{
-		// when fading
-		if (fadeAlpha == 255) {
-            blit(bmp_screen, screen, 0, 0, 0, 0, screen_x, screen_y);
-        } else {
-			BITMAP *temp = create_bitmap(game.screen_x, game.screen_y);
-			assert(temp != NULL);
-			clear(temp);
-			fblend_trans(bmp_screen, temp, 0, 0, fadeAlpha);
-			blit(temp, screen, 0, 0, 0, 0, screen_x, screen_y);
-			destroy_bitmap(temp);
+		if (fadeAction == eFadeAction::FADE_NONE) {
+  		// Not shaking and not fading.
+      vsync();
+      blit(bmp_screen, screen, 0, 0, 0, 0, screen_x, screen_y);
+    } else {
+      // Fading
+      assert(fadeAlpha >= kMinAlpha);
+      assert(fadeAlpha <= kMaxAlpha);
+      auto temp = std::unique_ptr<BITMAP, decltype(&destroy_bitmap)>(create_bitmap(game.screen_x, game.screen_y), destroy_bitmap);
```

Comment:
so, my c++ is not as proficient as yours.. let me try to unpack this:
the `std::unique_ptr` will trigger `destroy_bitmap` once the `temp` variable goes out of scope. And you instantiate it via the `create_bitmap` within its parentheses?
I do see `destroy_bitmap` twice in that line. 

Hmm. So I read at https://en.cppreference.com/w/cpp/memory/unique_ptr, but that is still a bit too much info for me. I also see `decltype` used, never saw that (did read https://en.cppreference.com/w/cpp/language/decltype though). I trust you on this 👍